Mix feature flags into fingerprint/metadata shorthash

nipunn1313 opened the pull request:
Since building dependencies results in different libraries
depending on the feature flags, I added the feature flags into
the short_hash.

This solves an issue when multiple crates share a target directory
or multiple targets share a common library with divergent feature
flag choice.

I'm not sure if this architecturally the best way to solve this problem, but I did confirm that this fixes the issue I was seeing. I can also add a test for this case if this code is taking the right approach (or if it would help illustrate the issue).

alexcrichton commented:
Thanks for the PR! Wanted to let you know that I hadn't forgotten about this and I've just been thinking about it recently. It looks like a test is currently failing but I think it's just because it's expecting a rebuild when one isn't actually happening, so not a hard failure. That to me though is proof enough that this works :)

Could you add a test though which actually runs the output to ensure everything gets properly linked? (just to make sure we don't accidentally use stale artifacts).

I think that this may start inflating the size of build directories over time a bit, but otherwise this seems pretty reasonable to me. I admit that I've run into this problem multiple times before as well and it seems nice to solve it once and for all.

nipunn1313 commented:
I also wanted to add a couple of tests for this case as well, so I'll take care of that and holler back when updated.

nipunn1313 commented:
Ok. I'm not confident this is quite doing the right thing for binaries. See my comments on the diff.

nipunn1313 commented on src/cargo/ops/cargo_rustc/context.rs:
```diff
@@ -358,9 +366,18 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
                 })
             } else {
                 None
+                /*
```
I left this code commented out to highlight a point.
I believe that when building a path dependency, this function returns None in order to not "pollute the suffix" of the executable or the library.

However, this means that when building across features, it appears to build to the same executable/library name.

It's correctly separating the .fingerprint, but not the actual built executable.

alexcrichton replied:
Hm yes I can see how this would indeed be a problem! For executables we could generate ones with long filenames and them link them into a location without the metadata, but that does weird things with dynamic libraries unfortunately, so we can't do that everywhere.

Short of building everything in an entirely new directory, we may have to just stomach recompiling these top level dependencies whenever various flags change.

nipunn1313 commented on src/cargo/ops/cargo_rustc/fingerprint.rs:
```diff
@@ -506,17 +506,24 @@ pub fn prepare_init(cx: &mut Context, unit: &Unit) -> CargoResult<()> {
+        dir.join(format!("feat-{}", feat_vec.join("-")))
+    } else {
+        dir
+    }
```
This is correctly separating out the fingerprints, but not the executables.

The executables/libraries are correctly separated as long as they are a full dep, but when they are a path dep, I think this is not working right.

nipunn1313 commented on tests/freshness.rs:
```diff
@@ -172,16 +172,192 @@ fn changing_features_is_ok() {
+            ftest  = []
+        "#)
+        .file("dep_crate/src/lib.rs", r#"
+            #[cfg(feature = "ftest")]
```
I would expect this test to fail because I conditionally compile this dep_crate to do different things on each branch depending on feature flags (and they're definitely building to one lib file with no suffixes)

```
nipunn-mbp:foo nipunn$ ls target/debug/
a         a.dSYM    b         b.dSYM    build     deps      examples  liba.rlib libb.rlib native
nipunn-mbp:foo nipunn$ ls target/debug/deps/
liba.rlib         libb.rlib         libdep_crate.rlib
```

nipunn1313 commented on tests/freshness.rs:
```diff
@@ -172,16 +172,192 @@ fn changing_features_is_ok() {
+[RUNNING] `[..]target/debug/b`
+"));
+
+    /* Build a/ package again. If we cache different feature dep builds correctly,
```
Given the commented out code in cargo, I'd expect this test to fail, especially given that its building dep_crate to the same path for both a/ and b/ (shared target dir), yet this is failing.

alexcrichton replied:
Hm so this test is passing?

If so, I think that may be due to how Cargo tracks where if one target is rebuilt it will unconditionally rebuild all transitive dependants. You can perhaps try to expose this through:

```
cd a
cargo build -p dep_crate
cargo run
```

That'll force Cargo to try to understand that the dependency was rebuilt, maybe? (not sure if that'd fail)

nipunn1313 commented:
Didn't forget about this! Just got a bit busy.
Would it be viable to also mix in the panic flag (eg panic=unwind vs
panic=abort)?

On Fri, Oct 7, 2016 at 3:29 PM Alex Crichton <notifications@github.com>
wrote:

> *@alexcrichton* commented on this pull request.
> ------------------------------
>
> In tests/freshness.rs <https://github.com/rust-lang/cargo/pull/3102>:
>
> > +[..]Compiling dep_crate v0.0.1 ([..])
> +[..]Compiling b v0.0.1 ([..])
> +[FINISHED] debug [unoptimized + debuginfo] target(s) in [..]
> +[RUNNING] `[..]target/debug/b`
> +"));
> +
> +    /* Rebuild b/ package. Should not recompile since we just built b */
> +    assert_that(p.cargo("run").cwd(p.root().join("b")),
> +                execs().with_status(0)
> +                       .with_stdout("ftest on")
> +                       .with_stderr("\
> +[FINISHED] debug [unoptimized + debuginfo] target(s) in [..]
> +[RUNNING] `[..]target/debug/b`
> +"));
> +
> +    /* Build a/ package again. If we cache different feature dep builds correctly,
>
> Hm so this test is passing?
>
> If so, I think that may be due to how Cargo tracks where if one target is
> rebuilt it will unconditionally rebuild all transitive dependants. You can
> perhaps try to expose this through:
>
> cd a
> cargo build -p dep_crate
> cargo run
>
> That'll force Cargo to try to understand that the dependency was rebuilt,
> maybe? (not sure if that'd fail)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/cargo/pull/3102>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPXo5tZwxqd6nrU2sylgcTlxZQaSmPIks5qxseugaJpZM4KAOeN>
> .
>

alexcrichton commented:
Yeah if we do take this route I think we'll want to hash in those options as well. I actually just had a chat with @brson as well and we're feeling like we may as well go whole hog on this and hash *everything* to ensure that bins are distinct. For now that basically means that we'd hash basically an entire `Profile` in Cargo which has all sorts of compiler configuration. Eventually we could also hash in information like the compiler we're compiling with.

nipunn1313 commented:
I went through and completed this PR.
- Fixed the test issue by cleaning just the parent package (to ensure the dep package rebuilt)
- Moved bin targets into the target/debug/deps/ directory with hash suffix.
- Hard link bin targets without hash suffix to both target/debug and target/debug/deps
- Hard link test targets with a "-test" suffix rather than a hash suffix.
- Fingerprint is based on pkg hash, but subdirectory based on target_fingerprint (same logic used for bin suffix generation).
- Fix existing tests which expect a non-suffix compile

Stuff I did not do, but could add in
- Mix in the entire unit.profile to the hash
- Build bin targets into some kind of target/debug/bin/ directory instead of deps?

Stuff I would appreciate some additional scrutiny
- Sufficient test coverage? How can we simplify the tests while maximizing test coverage?
- Complexity. I fear this diff has gotten a bit complex and some architectural things could make this a lot simpler, but I couldn't wrap my head around the whole codebase.

nipunn1313 commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -315,12 +321,36 @@ fn rustc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
+                     .chain_error(|| {
+                         human(format!("failed to link or copy `{}` to `{}`",
+                                       src.display(), dst.display()))
+                }));
```
Very interesting. This code is non-deterministically failing due to parallelism + the non-atomicity of the remove_file, hard_link, and  fs_copy on the test which is compiling two versions of a crate (use_a_spec_to_select).
Here is the ordering that causes a failure.

thread 1) foo-abcdef
thread 2) foo-123456

1) Checks if foo exists [no]
2) Checks if foo exists [no]
1) hard links foo(inode x) -> foo-abcdef(inode x)
2) Attempts to hard link foo -> foo-123456 (fails due to foo existing)
2) Attempts to copy foo-123456 -> foo (succeeds, overwriting inode x and both versions of foo).

Now foo, foo-123456, and foo-abcdef all have the same contents.
We need to somehow prevent either
1) Multiple versions of the same crate being compiled concurrently
2) Multiple versions attempting hard link concurrently.

nipunn1313 replied:
Perhaps linking within the deps/ directory is not necessary and we could try to eliminate this behavior of having a deps/libfoo.rlib. I will try that out.

nipunn1313 commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -335,8 +365,7 @@ fn rustc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
                 }
-                let dst = src_dir.parent().unwrap()
-                                 .join(src.file_name().unwrap());
+                let dst = src_dir.parent().unwrap().join(bin_name);
```
Might this code below (hard link when doing move_outputs_up) also be susceptible to the same bug? Perhaps not today, since we should only be moving up a top level package, but it seems like a fragility that might cause us to inadvertently add a bug in the future.

One possibility (striving for simplicity over performance) is to only copy and never link.

nipunn1313 commented:
Ok. I sucked it up and got my windows VM up and running with a developer environment. I narrowed the test issue down to this.

On windows, when producing a dylib with rustc, it produces four files in 
target/debug/deps:
- builder.dll
- builder.dll.exp
- builder.dll.lib
- builder.pdb

My code modifies this to have
target/debug/deps:
- builder-123456.dll
- builder-123456.dll.exp
- builder-123456.dll.lib
- builder-123456.pdb
target/debug
- builder.dll
My code only links the "target_filenames" to target/debug/ as specified from the context, which only includes "builder.dll" right now.

When setting as a dep for foo, it tries to link against builder.dll.lib (as seen here https://ci.appveyor.com/project/rust-lang-libs/cargo/build/1.0.44/job/8q6mclw54x5snn9j)

Thus, it fails to link since the target/debug/ directory only has the .dll and not the .dll.lib

Some possibilities, which I don't have enough context to select between =\
- Link all four from deps/ up to debug/ by adding to "target_filenames"
- Run rustc more than once for the ones which would link upward
- Don't put suffixes on dylibs because windows

alexcrichton commented:
> Fingerprint is based on pkg hash, but subdirectory based on target_fingerprint (same logic used for bin suffix generation).

Perhaps we could just base the whole fingerprint on the target fingerprint? E.g. include the whole hash here.

> Mix in the entire unit.profile to the hash

Yeah I think this'll give us the most bang for our buck, if it's easy to do let's go ahead and do it.

> Build bin targets into some kind of target/debug/bin/ directory instead of deps?

Nah the strategy you outlined above seems fine to me. I'm a little worried about dynamic libraries but I'll read over this and see what's what.

> Don't put suffixes on dylibs because windows

Oh geez that's a bummer. I guess this confirms my dylib worries above as well!

So OSX is also weird here where the dylib itself encodes the name it was compiled with. That way if you rename the dynamic library and link to the renamed copy, that binary won't work at runtime because it'll try to find the old name.

In that sense it's probably easiest to just not suffix dylibs with anything and completely recompile them more often (due to these issues).

alexcrichton commented on src/cargo/ops/cargo_rustc/context.rs:
```diff
@@ -309,49 +309,46 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
 
     /// Get the metadata for a target in a specific profile
     pub fn target_metadata(&self, unit: &Unit) -> Option<Metadata> {
```
I think this function should always return `Some` now, right? It may be useful to just return `Metadata` here and then deal with the consequences elsewhere as they arise.

alexcrichton commented on src/cargo/ops/cargo_rustc/context.rs:
```diff
@@ -360,19 +357,30 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
+            unit.target.crate_name()
+        };
+        if unit.profile.test {
+            bin + "-test"
```
Hm I'm trying to recall why something like this would be required but I'm drawing a blank. Could you remind me why this is needed here?

nipunn1313 replied:
Yeah. This was needed to separate test binaries from true binaries. Previously they were separated by metadata hash. The move-targets-up code was linking them on top of each other in this code. One option was to replicate the -{metadata} behavior, but I thought it simpler to just do "-test" instead now that there was a clear way of separating them.

alexcrichton commented on src/cargo/ops/cargo_rustc/fingerprint.rs:
```diff
@@ -82,8 +83,13 @@ pub fn prepare_target<'a, 'cfg>(cx: &mut Context<'a, 'cfg>,
     } else {
-        for (filename, _) in try!(cx.target_filenames(unit)) {
-            missing_outputs |= fs::metadata(root.join(filename)).is_err();
+        let move_outputs_up = unit.pkg.package_id() == &cx.current_package || unit.target.is_bin();
```
This looks like it may want to be extracted to a method on `Context`?

alexcrichton commented on tests/build.rs:
```diff
@@ -1790,13 +1790,15 @@ fn example_bin_same_name() {
         .unwrap();
 
     assert_that(&p.bin("foo"), is_not(existing_file()));
+    assert_that(&p.bin("foo-test"), existing_file());
```
Hm ok so this may be where the `-test` is showing up.

This looks like a somewhat orthogonal change, however (having predictably named test artifacts). Perhaps we could leave that to a future PR?

nipunn1313 commented:
Sure I can take another look at it. Comments seem to make sense.

nipunn1313 commented:
Hey @alexcrichton. I noticed that the appveyor is getting backed up for hours. Enabling this feature may help reduce the size of the queue when something is under rapid development. Alternately, could upgrade the appveyor account for additional parallelism, but honestly, rolling builds will probably be good enough.

https://www.appveyor.com/docs/build-configuration/#rolling-builds

nipunn1313 commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -319,36 +323,30 @@ fn rustc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
+                    human(format!("failed to remove: {}", dst.display()))
                 }));
             }
+            try!(fs::hard_link(&src, &dst)
```
Do we have to pop this linking step out into a separate task? I'm worried that the task won't re-run under all conditions. For example, if there are two binaries built with different features, it would compile to
deps/bin-abc123
deps/bin-456789
If you then switched back to using the first binary, the fingerprint would indicate that you're all set. Since the linking is happening as part of the rustc stage, there's no way to re-link without recompiling. Any suggestion on how to get around that?

For now, I will put together a test that fails this condition and see if I can get some code to trigger a recompile of the top level package.

alexcrichton replied:
Hm, a very good point! Definitely something we'll want to handle. This makes me vaguely wonder how this works today...

IIRC, each unit of work always has something run. There's a unit of work to run if it's dirty and a different one if it's "fresh" (doesn't need to be rerun). The thinking is that "fresh" work is way faster than dirty work. 

So I think we can shove this into also happening during the fresh stages to cover this use case?

nipunn1313 replied:
This works today because the fingerprint filename is the same for both deps. Thus if you change the features and try to recompile, it will trigger as dirty. I changed it to have different fingerprint filename for each of these variants so when switching back and forth we keep it working right.

bors commented:
:umbrella: The latest upstream changes (presumably #3257) made this pull request unmergeable. Please resolve the merge conflicts.

nipunn1313 commented:
FWIW this has ended up becoming quite the large diff. It could feasibly be broken into two parts
1) Refactor target_filenames into returning the 3-tuple with link_dst without any semantic changes
2) Make semantic changes to target_dst, link_dst, fingerprint, and dylib.

If you think it is worthwhile, I'm happy to do it, since it will definitely shorten the diffs. Just want to make sure part 1 is in line with your view before I go for it.

alexcrichton commented:
Ah nah this is fine in terms of size, I personally prefer to consume everything all at once anyway to see how it all fits together.

alexcrichton commented on src/cargo/ops/cargo_rustc/context.rs:
```diff
@@ -309,49 +309,55 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
+        // No metadata for dylibs because of a couple issues
+        // - OSX encodes the dylib name in the executable
+        // - Windows rustc multiple files of which we can't easily link all of them
+        if !unit.profile.test && unit.target.is_dylib() {
```
Ah so one thing I just remembered though is that we don't want to unconditionally return `None` here in the case of an upstream dylib. That is, if a transitive dependency is a dylib, we'll want metadata mixed in there to resolve name conflicts.

Additionally, the `__CARGO_DEFAULT_LIB_METADATA` below I think still needs to be factored in here as well. That's used when building the standard library itself to ensure that we throw a hash in there even though it's a dylib and even though it's a local path dependency.

nipunn1313 replied:
What are the conditions we want? We want to check if the target is the current_package AND is a dylib?

alexcrichton replied:
> How to handle the upstream dylib case

I think the conditions we're going to want here are:

```rust
if !unit.profile.test && 
   unit.target.is_dylib() && 
   unit.pkg.package_id().source_id().is_path() &&
   magical_env_var_not_set() {
    return None
}
```

That is, we want metadata on all tests, we want metadata on everything that's not a dylib, we want metdata on all upstream deps (non-path deps), and finally we have this magical env var thing.

Note that "upstream" == "not a path dep" because the current package changes when you move around a workspace. A path dep is basically a heuristic, but it works fine for us.

So for the magical env var, the problem is that if we *didn't* have that clause in the condition then when we `cargo build` the standard library we get a bunch of files that are literally `libstd.so`, `librustc.so`, etc. Now we don't want this because it means when you install Rust on your system you'd get `/usr/lib/libstd.so`. That's a little too "official"  for us to take! To solve this problem we've always added some form of hash to the libraries, so you'd instead install `/usr/lib/libstd-xxxxxxxx.so`. 

So basically when we use Cargo to build the standard library we want to unconditionally insert all these hashes, so the env var just says "please insert hashes everywhere".

Now if you're curious why we don't *always* install the hash, that's also interesting. The main idea is that if I have a project called "foo" and I built a dylib, we want to generate the predictable filename `target/debug/libfoo.so`. (no hash). Now for this to actually work correctly, however, we can't take the same trick we're doing with staticlibs and executables where we just hard link the file to the correct name. Platforms like OSX use the generate file name (`target/debug/deps/libfoo-xxxxxx.so`) at runtime for various reasons, so we need to generate the file as literally `target/debug/deps/libfoo.so` for now. We can fix this with linker options but rustc doesn't expose that yet (and is a problem for another day!)

Does that all make sense?

alexcrichton commented on src/cargo/ops/cargo_rustc/context.rs:
```diff
@@ -309,49 +309,55 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
+            }
+            m.mix(unit.profile);
+            m
+        };
```
In retrospect I'm unfortunately finding this kinda confusing (why we'd choose one or the other). I think nowadays it's probably best to just defer all of this until right here. That is, we should remove the `Metadata` inside of `Target` and calculate everything here (doing all the special casing that's happening during the construction. 

We seem to basically be doing all the special casing here anyway, so the initial construction in `toml.rs` isn't buying us much anyway.

alexcrichton replied:
(note to myself seeing this comment in the future, we're punting on this to a future refactor)

alexcrichton commented on src/cargo/ops/cargo_rustc/context.rs:
```diff
@@ -429,6 +476,7 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
                    support any of the output crate types",
                   unit.pkg, self.target_triple());
         }
+        info!("Target filenames: {:?}", ret);
```
👍

alexcrichton commented on src/cargo/ops/cargo_rustc/fingerprint.rs:
```diff
@@ -82,8 +82,11 @@ pub fn prepare_target<'a, 'cfg>(cx: &mut Context<'a, 'cfg>,
-        for (filename, _) in try!(cx.target_filenames(unit)) {
-            missing_outputs |= fs::metadata(root.join(filename)).is_err();
+        for (src, link_dst, _) in try!(cx.target_filenames(unit)) {
+            missing_outputs |= fs::metadata(&src).is_err();
```
This was actually kept around from literally Rust 1.0.0, but nowadays it's fine to change this to `!src.exists()`

alexcrichton commented on src/cargo/ops/cargo_rustc/context.rs:
```diff
@@ -360,19 +366,57 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
+        }
+    }
+
+    pub fn link_stem(&self, unit: &Unit) -> Option<(PathBuf, String)> {
```
Could a doc block be added here explaining what this is doing? Still working through what each returned value represents...

nipunn1313 replied:
Sure thing.

alexcrichton commented on src/cargo/ops/cargo_rustc/layout.rs:
```diff
@@ -175,7 +175,7 @@ impl<'a> LayoutProxy<'a> {
             self.deps().to_path_buf()
         } else {
-            self.root().to_path_buf()
+            self.deps().to_path_buf()
```
You can probably collapse this with the previous `else if` now

alexcrichton commented on src/cargo/ops/cargo_rustc/fingerprint.rs:
```diff
@@ -650,7 +653,13 @@ fn mtime_if_fresh<I>(output: &Path, paths: I) -> Option<FileTime>
+    // file_stem (which can include a suffix)
+    let file_stem = cx.link_stem(unit)
+        .map(|(_path, stem)| stem)
+        .unwrap_or_else(|| cx.file_stem(unit));
```
Whoa! This is quite different than before (no nested fingerprint directories).

If feel like this is where all the magic, and I guess that's because `file_stem` includes a metadata hash in it?

nipunn1313 replied:
Yep. Given your suggestion to also link in the Fresh step means that we can get rid of the "link_stem" logic, since using the file_stem is safe (different fingerprint for each metadata).

alexcrichton commented on src/cargo/ops/cargo_rustc/context.rs:
```diff
@@ -360,19 +366,57 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
+        // it was compiled into something like `example/` or `doc/` then
+        // we don't want to link it up.
+        if src_dir.ends_with("deps") {
+            // Don't lift up library dependencies
```
I *think* that if all tests pass this is fine to just have the `else` part. I think this came from way down below in `cargo_rustc/mod.rs` but I can't quite recall exactly what this is doing. Although the doc block on this function may help me refresh my memory :)

nipunn1313 replied:
I think this if else prevents us from linking up dependent libs.

Eg if crate foo depends on something else (say petgraph)
This if else prevents us from linking `target/debug/deps/petgraph-123456.rlib` to `target/debug/petgraph.rlib`. There's no point in linking up deps, though it doesn't hurt.

alexcrichton replied:
Oh ok, so the test would be asserting that `target/debug/libpetgraph.rlib` *doesn't* exist. In that sense this behavior definitely looks correct to me, and if you want to add a test for that, sounds good to me!

nipunn1313 replied:
Added a couple lines to an existing test. No prob!

alexcrichton commented:
> I noticed that the appveyor is getting backed up for hours

Hm yeah I wouldn't worry too much about that though, this happens occasionally but it tends to not be too much of a problem to land PRs at least (most of the time)

alexcrichton commented on tests/freshness.rs:
```diff
@@ -172,6 +172,13 @@ fn changing_features_is_ok() {
+           not separate from the "compile" step. Packages which link
+           to top level binaries eg (deps/foo-abc123 -> foo) are forced
+           to do an extra recompile here.
+    */
```
If we modify the "symlink" step to also happen in the "fresh" unit of work I think we can solve this, right?

alexcrichton commented on tests/freshness.rs:
```diff
@@ -182,6 +189,264 @@ fn changing_features_is_ok() {
+           not separate from the "compile" step. Packages which link
+           to top level binaries eg (deps/foo-abc123 -> foo) are forced
+           to do an extra recompile here.
+    */
```
In theory nothing in this test needs to be symlinked though, other than the rlib? Or is that the one causing the problem?

nipunn1313 replied:
Both parts of this test requires symlink. Currently, this test is passing because it's queueing a recompile (which is doing the relink), but with your symlink-in-fresh idea, I was able to get this to work with no recompiles! Hooray.

nipunn1313 commented:
It's no problem! The work isn't that taxing. I think it's just getting tricky enough that it's tougher to reason about correctness. Having multiple eyes is really helpful. I'm also pretty excited about this change because it causes us a fair deal of full recompiles.

Moving the symlinking to fresh is a great idea. That will probably get us the last bit.

alexcrichton commented:
👍 

Sounds good to me! This has been simplified to the point that I'm quite confident in it, so as long as the tests are green then I'm happy to land at any time.

nipunn1313 commented on tests/build.rs:
```diff
@@ -1789,13 +1789,15 @@ fn example_bin_same_name() {
                    .unwrap();
 
     assert_that(&p.bin("foo"), is_not(existing_file()));
+    // We expect a file of the form bin/foo-{metadata_hash}
```
Nope this is intentional. `bin/foo` does not exist, but `bin/foo-{metadatahash}` and `bin/examples/foo` both exist. The test is making sure that both files are created, but it can't test for `bin/foo-{metadata hash}` effectively.

alexcrichton replied:
Oh I see, this comment was intended for the previous line, not the next, carry on!

nipunn1313 commented on tests/freshness.rs:
```diff
@@ -182,6 +189,264 @@ fn changing_features_is_ok() {
+[FINISHED] debug [unoptimized + debuginfo] target(s) in [..]
+[RUNNING] `[..]target[/]debug[/]b[EXE]`
+"));
+}
```
I think the cargo clean is necessary because we're testing the freshness of the dep_crate. We need to somehow dirty "a" before recompiling "a" or it's always going to be clean. We're specifically testing the situation where "a" is dirty but "dep_crate" is clean.

We could alternately test this by actually dirtying one of the source files if that's better, but I think this is fine.

alexcrichton replied:
Oh I thought this was a test case where currently when you alternate `cargo build` in projects with different activated feature sets it always recompiles everything. Here though it shouldn't recompile anything if nothing was changed, right? Perhaps this is just a separate test I'm thinking of?

alexcrichton commented:
> Decided to punt on the refactor to remove Metadata

👍 

I'll reply to others inline. Assuming I can figure out how as GitHub seems to not want to let me...

alexcrichton commented:
Ok everything looks great to me! I think with the updates to dylib naming I'm good to r+

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 3e77640 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 3e776404384efd03d6f8e8dfd44494c0ec33e7e9 -->

bors commented:
:hourglass: Testing commit 3e77640 with merge 25f87a6...

bors commented:
:broken_heart: Test failed - [cargo-cross-linux](https://buildbot.rust-lang.org/builders/cargo-cross-linux/builds/365)

alexcrichton commented:
Spurious failure, I'll work on getting this landed

alexcrichton commented:
@bors: retry

On Mon, Nov 14, 2016 at 10:49 AM bors <notifications@github.com> wrote:

> 💔 Test failed - cargo-cross-linux
> <https://buildbot.rust-lang.org/builders/cargo-cross-linux/builds/365>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/cargo/pull/3102#issuecomment-260390787>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95Hd27f9pA0qhUa8XpoPYtCfr3EfFks5q-JEegaJpZM4KAOeN>
> .
>

bors commented:
:hourglass: Testing commit 3e77640 with merge 9181681...

bors commented:
:broken_heart: Test failed - [cargo-win-msvc-64](https://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/759)

alexcrichton commented:
@bors: retry

On Mon, Nov 14, 2016 at 7:35 PM bors <notifications@github.com> wrote:

> 💔 Test failed - cargo-win-msvc-64
> <https://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/759>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/cargo/pull/3102#issuecomment-260520557>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95CaQ57m42HoBrxcD1H4gZPXUVfS_ks5q-QxjgaJpZM4KAOeN>
> .
>

bors commented:
:zap: Previous build results for [cargo-cross-linux](https://buildbot.rust-lang.org/builders/cargo-cross-linux/builds/370), [cargo-linux-32](https://buildbot.rust-lang.org/builders/cargo-linux-32/builds/1930), [cargo-linux-64](https://buildbot.rust-lang.org/builders/cargo-linux-64/builds/1916), [cargo-mac-32](https://buildbot.rust-lang.org/builders/cargo-mac-32/builds/1922), [cargo-mac-64](https://buildbot.rust-lang.org/builders/cargo-mac-64/builds/1913) are reusable. Rebuilding only [cargo-win-gnu-32](https://buildbot.rust-lang.org/builders/cargo-win-gnu-32/builds/809), [cargo-win-gnu-64](https://buildbot.rust-lang.org/builders/cargo-win-gnu-64/builds/811), [cargo-win-msvc-32](https://buildbot.rust-lang.org/builders/cargo-win-msvc-32/builds/728), [cargo-win-msvc-64](https://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/759)...

bors commented:
:broken_heart: Test failed - [cargo-win-gnu-32](https://buildbot.rust-lang.org/builders/cargo-win-gnu-32/builds/810)

alexcrichton commented:
@bors: retry

On Mon, Nov 14, 2016 at 9:48 PM bors <notifications@github.com> wrote:

> 💔 Test failed - cargo-win-gnu-32
> <https://buildbot.rust-lang.org/builders/cargo-win-gnu-32/builds/810>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/cargo/pull/3102#issuecomment-260540096>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95B18FDuOT3P3p3sGeLSpJYPGzj-Bks5q-SucgaJpZM4KAOeN>
> .
>

bors commented:
:zap: Previous build results for [cargo-cross-linux](https://buildbot.rust-lang.org/builders/cargo-cross-linux/builds/370), [cargo-linux-32](https://buildbot.rust-lang.org/builders/cargo-linux-32/builds/1930), [cargo-linux-64](https://buildbot.rust-lang.org/builders/cargo-linux-64/builds/1916), [cargo-mac-32](https://buildbot.rust-lang.org/builders/cargo-mac-32/builds/1922), [cargo-mac-64](https://buildbot.rust-lang.org/builders/cargo-mac-64/builds/1913) are reusable. Rebuilding only [cargo-win-gnu-32](https://buildbot.rust-lang.org/builders/cargo-win-gnu-32/builds/810), [cargo-win-gnu-64](https://buildbot.rust-lang.org/builders/cargo-win-gnu-64/builds/812), [cargo-win-msvc-32](https://buildbot.rust-lang.org/builders/cargo-win-msvc-32/builds/729), [cargo-win-msvc-64](https://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/760)...

bors commented:
:broken_heart: Test failed - [cargo-win-msvc-64](https://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/761)

alexcrichton commented:
@bors: retry

Hm I think I need to disable that test on Windows, it seems inherently racy...

bors commented:
:zap: Previous build results for [cargo-cross-linux](https://buildbot.rust-lang.org/builders/cargo-cross-linux/builds/370), [cargo-linux-32](https://buildbot.rust-lang.org/builders/cargo-linux-32/builds/1930), [cargo-linux-64](https://buildbot.rust-lang.org/builders/cargo-linux-64/builds/1916), [cargo-mac-32](https://buildbot.rust-lang.org/builders/cargo-mac-32/builds/1922), [cargo-mac-64](https://buildbot.rust-lang.org/builders/cargo-mac-64/builds/1913) are reusable. Rebuilding only [cargo-win-gnu-32](https://buildbot.rust-lang.org/builders/cargo-win-gnu-32/builds/811), [cargo-win-gnu-64](https://buildbot.rust-lang.org/builders/cargo-win-gnu-64/builds/813), [cargo-win-msvc-32](https://buildbot.rust-lang.org/builders/cargo-win-msvc-32/builds/730), [cargo-win-msvc-64](https://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/761)...

nipunn1313 commented:
Oh yeah. Windows often has different resource reclamation on process unexpected exit. I've seen this sort of thing happen before resulting in an error reopening the file for a little bit. The issue is often transient, but there may be a way to force it so the test isn't spurious.

bors commented:
:broken_heart: Test failed - [cargo-win-msvc-64](https://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/762)

nipunn1313 commented:
Perhaps since my change is dealing with hard links a lot more often, it's increasing the flake probability? Just a total guess.

nipunn1313 commented:
Opened #3293 to make an attempt at fixing the flake. I know windows can be finicky about when resources are reclaimed (especially around multiple processes accessing the same files)

alexcrichton commented:
@bors: retry

[left a comment](https://github.com/rust-lang/cargo/pull/3293#issuecomment-260666020) but in the meantime may as well try to keep landing this.

I guess yeah behavior changes here but tbh this sort of stuff just happens, for a few hours the bots just feel like not landing anything...

bors commented:
:hourglass: Testing commit 3e77640 with merge 605dae6...

bors commented:
:broken_heart: Test failed - [cargo-win-msvc-64](https://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/764)

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -383,6 +355,44 @@ fn rustc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
+            debug!("linking {} to {}", src.display(), dst.display());
+            if dst.exists() {
+                try!(fs::remove_file(&dst).chain_error(|| {
+                    human(format!("failed to remove: {}", dst.display()))
```
Oh so one reason I think that the test is basically deterministically failing is that this happens both in fresh and dirty modes now. That is, with a clean build we'd still be doing fs ops earlier in the build and presumably that's now closing the race window to expose the existing bug on Windows?

nipunn1313 replied:
That's an interesting hypothesis. I suppose that when running cargo build twice concurrently as in that test, the second build used to hit "fresh" without having to try to access the output. (It could still race on accessing the fingerprint files, but less likely). Now fresh also accesses the output.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 235ebde has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 235ebde7d07cceec39bb75c783d7fa66684c8ba2 -->

bors commented:
:lock: Merge conflict

bors commented:
:umbrella: The latest upstream changes (presumably #3283) made this pull request unmergeable. Please resolve the merge conflicts.

nipunn1313 commented:
Ah this is mostly due to try! converting to ? syntax. I can fix this.

alexcrichton commented:
I swear github literally flips a coin when a branch is updated as to whether an email notification is sent for that. I sometimes get them and sometimes don't...

@bors: r+

bors commented:
:pushpin: Commit d7977f8 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton d7977f881f6017ff96b9db889d830d12b788a129 -->

bors commented:
:hourglass: Testing commit d7977f8 with merge 1cf3cba...

bors commented:
:broken_heart: Test failed - [cargo-win-gnu-32](https://buildbot.rust-lang.org/builders/cargo-win-gnu-32/builds/817)

alexcrichton commented on tests/concurrent.rs:
```diff
@@ -343,7 +343,9 @@ fn same_project() {
 // next Cargo to come in will take over cleanly.
+// older win versions don't support job objects, so skip test there
 #[test]
+#[cfg_attr(target_env = "msvc", ignore)]
```
Ah I guess this needs to be `windows`, not just MSVC

nipunn1313 commented:
That seems like it'll cause you to lose test coverage on windows for some code that's specific to windows. But here you go!

alexcrichton commented:
Ah that's ok, as soon as we ditch buildbot I'll re-enable the test

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 2af9e1f has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 2af9e1fac580faaf8ae930c2981f4b2aa2740b99 -->

bors commented:
:hourglass: Testing commit 2af9e1f with merge e82ad8a...

bors commented:
:broken_heart: Test failed - [cargo-win-msvc-32](https://buildbot.rust-lang.org/builders/cargo-win-msvc-32/builds/737)

nipunn1313 commented:
Whoops. Said target_env = windows instead of target_os = windows. Fixing.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 8e22eca has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 8e22eca9c8f013aad4ca95b7870243b4eb8ac8c6 -->

bors commented:
:hourglass: Testing commit 8e22eca with merge 1877f59...

bors commented:
:sunny: Test successful - [cargo-cross-linux](https://buildbot.rust-lang.org/builders/cargo-cross-linux/builds/377), [cargo-linux-32](https://buildbot.rust-lang.org/builders/cargo-linux-32/builds/1937), [cargo-linux-64](https://buildbot.rust-lang.org/builders/cargo-linux-64/builds/1923), [cargo-mac-32](https://buildbot.rust-lang.org/builders/cargo-mac-32/builds/1929), [cargo-mac-64](https://buildbot.rust-lang.org/builders/cargo-mac-64/builds/1920), [cargo-win-gnu-32](https://buildbot.rust-lang.org/builders/cargo-win-gnu-32/builds/819), [cargo-win-gnu-64](https://buildbot.rust-lang.org/builders/cargo-win-gnu-64/builds/821), [cargo-win-msvc-32](https://buildbot.rust-lang.org/builders/cargo-win-msvc-32/builds/738), [cargo-win-msvc-64](https://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/769)
Approved by: alexcrichton
Pushing 1877f59d6b2cb057f7ef6c6b34b926fd96a683c1 to master...
